File: ios/LoginViewController.m
// Copyright 2018 David Sansome
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#import "LoginViewController.h"

#import "Tsurukame-Swift.h"

#import <PromiseKit/PromiseKit.h>

NS_ASSUME_NONNULL_BEGIN

NSNotificationName kLogoutNotification = @"kLogoutNotification";

static NSString *const kPrivacyPolicyURL =
    @"https://github.com/davidsansome/tsurukame/wiki/Privacy-Policy";

@interface LoginViewController () <UITextFieldDelegate>

@property(weak, nonatomic) IBOutlet UILabel *signInLabel;
@property(weak, nonatomic) IBOutlet UITextField *usernameField;
@property(weak, nonatomic) IBOutlet UITextField *passwordField;
@property(weak, nonatomic) IBOutlet UIButton *signInButton;
@property(weak, nonatomic) IBOutlet UILabel *privacyPolicyLabel;
@property(weak, nonatomic) IBOutlet UIButton *privacyPolicyButton;
@property(weak, nonatomic) IBOutlet UIView *activityIndicatorOverlay;
@property(weak, nonatomic) IBOutlet UIActivityIndicatorView *activityIndicator;

@end

@implementation LoginViewController

- (void)viewDidLoad {
  [super viewDidLoad];

  [TKMStyle addShadowToView:_signInLabel offset:0.f opacity:1.f radius:5.f];
  [TKMStyle addShadowToView:_privacyPolicyLabel offset:0.f opacity:1.f radius:2.f];
  [TKMStyle addShadowToView:_privacyPolicyButton offset:0.f opacity:1.f radius:2.f];

  if (_forcedUsername.length) {
    _usernameField.text = _forcedUsername;
    _usernameField.enabled = NO;
  }

  _usernameField.delegate = self;
  _passwordField.delegate = self;

  [_usernameField addTarget:self
                     action:@selector(textFieldDidChange:)
           forControlEvents:UIControlEventEditingChanged];
  [_passwordField addTarget:self
                     action:@selector(textFieldDidChange:)
           forControlEvents:UIControlEventEditingChanged];
  [self textFieldDidChange:_usernameField];
}

- (void)viewWillAppear:(BOOL)animated {
  [super viewWillAppear:animated];
  self.navigationController.navigationBarHidden = YES;
}

#pragma mark - UITextFieldDelegate

- (BOOL)textFieldShouldReturn:(UITextField *)textField {
  if (textField == _usernameField) {
    [_passwordField becomeFirstResponder];
  } else if (textField == _passwordField) {
    [self didTapSignInButton:self];
  }
  return YES;
}

- (void)textFieldDidChange:(UITextField *)textField {
  bool enabled = _usernameField.text.length != 0 && _passwordField.text.length != 0;
  _signInButton.enabled = enabled;
  _signInButton.backgroundColor = enabled ? TKMStyle.radicalColor2 : TKMStyleColor.grey33;
}

#pragma mark - Sign In flow

- (IBAction)didTapSignInButton:(id)sender {
  if (!_signInButton.enabled) {
    return;
  }
  [self showActivityIndicatorOverlay:true];

  WaniKaniWebClient *client = [[WaniKaniWebClient alloc] init];
  AnyPromise *promise = [client loginWithUsername:_usernameField.text password:_passwordField.text];
  promise
      .then(^(LoginResult *result) {
        NSLog(@"Login success!");
        Settings.userCookie = result.cookie;
        Settings.userApiToken = result.apiToken;
        Settings.userEmailAddress = result.emailAddress;

        [_delegate loginComplete];
      })
      .catch(^(NSError *error) {
        NSString *message;
        if ([error.domain isEqual:WaniKaniWebClientErrorDomain]) {
          message = [WaniKaniWebClient errorDescription:error.code];
        } else {
          message = error.localizedDescription;
        }
        [self showLoginError:message];
      });
}

#pragma mark - Errors and competion

- (void)showLoginError:(NSString *)message {
  dispatch_async(dispatch_get_main_queue(), ^{
    UIAlertController *c = [UIAlertController alertControllerWithTitle:@"Error"
                                                               message:message
                                                        preferredStyle:UIAlertControllerStyleAlert];
    [c addAction:[UIAlertAction actionWithTitle:@"Close"
                                          style:UIAlertActionStyleCancel
                                        handler:nil]];
    [self presentViewController:c animated:YES completion:nil];
    [self showActivityIndicatorOverlay:false];
  });
}

- (void)showActivityIndicatorOverlay:(bool)visible {
  [self.view endEditing:YES];
  _activityIndicatorOverlay.hidden = !visible;
  _activityIndicator.hidden = !visible;
  if (visible) {
    [_activityIndicator startAnimating];
  } else {
    [_activityIndicator stopAnimating];
  }
}

#pragma mark - Privacy policy

- (IBAction)didTapPrivacyPolicyButton:(id)sender {
  NSURL *url = [NSURL URLWithString:kPrivacyPolicyURL];
  NSDictionary<NSString *, id> *options = [NSDictionary dictionary];
  [[UIApplication sharedApplication] openURL:url options:options completionHandler:nil];
}

@end

NS_ASSUME_NONNULL_END

File: ios/LoginViewController.swift
// Copyright 2024 David Sansome
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import Foundation
import PromiseKit
import WaniKaniAPI

extension Notification.Name {
  static let logout = Notification.Name("kLogoutNotification")
}

private let kPrivacyPolicyURL = "https://github.com/davidsansome/tsurukame/wiki/Privacy-Policy"

protocol LoginViewControllerDelegate: AnyObject {
  func loginComplete()
}

class LoginViewController: UIViewController, UITextFieldDelegate {
  weak var delegate: LoginViewControllerDelegate?
  var forcedEmail: String?

  @IBOutlet private var signInLabel: UILabel!
  @IBOutlet private var emailField: UITextField!
  @IBOutlet private var passwordField: UITextField!
  @IBOutlet private var signInButton: UIButton!
  @IBOutlet private var apiTokenStack: UIStackView!
  @IBOutlet private var apiTokenField: UITextField!
  @IBOutlet private var pasteButton: UIButton!
  @IBOutlet private var swapLoginMethodsButton: UIButton!
  @IBOutlet private var privacyPolicyLabel: UILabel!
  @IBOutlet private var privacyPolicyButton: UIButton!
  @IBOutlet private var activityIndicatorOverlay: UIView!
  @IBOutlet private var activityIndicator: UIActivityIndicatorView!

  override func viewDidLoad() {
    super.viewDidLoad()

    if #available(iOS 13.0, *) {
      overrideUserInterfaceStyle = .light
    }

    TKMStyle.addShadowToView(signInLabel, offset: 0, opacity: 1, radius: 5)
    TKMStyle.addShadowToView(privacyPolicyLabel, offset: 0, opacity: 1, radius: 2)
    TKMStyle.addShadowToView(privacyPolicyButton, offset: 0, opacity: 1, radius: 2)
    TKMStyle.addShadowToView(swapLoginMethodsButton, offset: 0, opacity: 1, radius: 2)
    TKMStyle.addShadowToView(pasteButton, offset: 0, opacity: 1, radius: 5)

    if let forcedEmail = forcedEmail {
      emailField.text = forcedEmail
      emailField.isEnabled = false
    }

    emailField.delegate = self
    passwordField.delegate = self
    apiTokenField.delegate = self

    emailField.addTarget(self, action: #selector(textFieldDidChange), for: .editingChanged)
    passwordField.addTarget(self, action: #selector(textFieldDidChange), for: .editingChanged)
    apiTokenField.addTarget(self, action: #selector(textFieldDidChange), for: .editingChanged)
    textFieldDidChange(emailField)
  }

  override func viewWillAppear(_ animated: Bool) {
    super.viewWillAppear(animated)
    navigationController?.isNavigationBarHidden = true
  }

  // MARK: - UITextFieldDelegate

  func textFieldShouldReturn(_ textField: UITextField) -> Bool {
    if textField == emailField {
      passwordField.becomeFirstResponder()
    } else if textField == passwordField || textField == apiTokenField {
      didTapSignInButton()
    }
    return true
  }

  @objc func textFieldDidChange(_: UITextField) {
    updateSignInButtonState()
  }

  private func updateSignInButtonState() {
    var enabled = false
    if emailField.isHidden {
      enabled = !(apiTokenField.text?.isEmpty ?? true)
    } else {
      enabled = !(emailField.text?.isEmpty ?? true) && !(passwordField.text?.isEmpty ?? true)
    }
    signInButton.isEnabled = enabled
    signInButton.backgroundColor = enabled ? TKMStyle.radicalColor2 : TKMStyle.Color.grey33
  }

  // MARK: - Sign In flow

  @IBAction func didTapSignInButton() {
    if !signInButton.isEnabled {
      return
    }
    showActivityIndicatorOverlay(true)

    if !emailField.isHidden {
      let client = WaniKaniWebClient()
      let promise = client.login(email: emailField.text!, password: passwordField.text!)
      promise.done { result in
        NSLog("Login success!")
        Settings.userCookie = result.cookie
        Settings.userApiToken = result.apiToken
        Settings.userEmailAddress = self.emailField.text!

        self.delegate?.loginComplete()
      }.catch { error in
        self.showLoginError(error.localizedDescription)
      }
    } else {
      let token = apiTokenField.text!
      let apiClient = WaniKaniAPIClient(apiToken: token)
      let promise = apiClient.user(progress: Progress())
      promise.done { user in
        NSLog("Login success! User is at level: \(user.currentLevel)")
        Settings.userCookie = ""
        Settings.userApiToken = token
        Settings.userEmailAddress = ""
        self.delegate?.loginComplete()
      }.catch { err in
        self.showLoginError("Unable to login with API token! (\(err.localizedDescription))")
      }
    }
  }

  @IBAction func didTapSwapLoginMethods() {
    UIView.animate(withDuration: 0.1,
                   delay: 0.0,
                   options: [.curveLinear],
                   animations: {
                     let animatingInUserPass = self.emailField.isHidden
                     self.emailField.isHidden = !animatingInUserPass
                     self.passwordField.isHidden = !animatingInUserPass

                     self.apiTokenStack.isHidden = animatingInUserPass

                     self.emailField.alpha = animatingInUserPass ? 1 : 0
                     self.passwordField.alpha = animatingInUserPass ? 1 : 0
                     self.apiTokenStack.alpha = animatingInUserPass ? 0 : 1
                   }) { _ in
      let title = self.emailField.isHidden ? "Use email and password" : "Use API token"
      self.swapLoginMethodsButton.setTitle(title, for: .normal)
      self.updateSignInButtonState()
    }
  }

  @IBAction func didTapPasteButton(_: Any) {
    if let text = UIPasteboard.general.string {
      apiTokenField.text = text
      updateSignInButtonState()
    }
  }

  // MARK: - Errors and competion

  func showLoginError(_ message: String) {
    DispatchQueue.main.async {
      let c = UIAlertController(title: "Error", message: message, preferredStyle: .alert)
      c.addAction(UIAlertAction(title: "Close", style: .cancel, handler: nil))
      self.present(c, animated: true, completion: nil)
      self.showActivityIndicatorOverlay(false)
    }
  }

  func showActivityIndicatorOverlay(_ visible: Bool) {
    view.endEditing(true)
    activityIndicatorOverlay.isHidden = !visible
    activityIndicator.isHidden = !visible
    if visible {
      activityIndicator.startAnimating()
    } else {
      activityIndicator.stopAnimating()
    }
  }

  // MARK: - Privacy policy

  @IBAction func didTapPrivacyPolicyButton() {
    let url = URL(string: kPrivacyPolicyURL)!
    UIApplication.shared.open(url, options: [:], completionHandler: nil)
  }
}
